### ISSUE TITLE
[LoRA Phase 10.5] Implement Multi-GPU Training Support

### ISSUE THREAD

**Issue by makr-code** [makr-code/ThemisDB#565]
## 📋 Description

**Priority**: Phase 10.5 - Priority #5 (Scalability)  
**Prerequisites**: Phases 1-9 complete, mixed precision (Phase 10.4) recommended  
**Estimated Effort**: 4 weeks  
**Status Document**: `LORA_GPU_PHASE10_PLAN.md`

Implement data parallelism for multi-GPU LoRA training. Distribute training across multiple GPUs with gradient synchronization for near-linear scaling. Supports NCCL (NVIDIA), RCCL (AMD), and custom all-reduce implementations.

## 🎯 Goals

- [ ] Data parallelism across multiple GPUs
- [ ] Near-linear scaling (4 GPUs → 3.5-3.8x speedup)
- [ ] Gradient synchronization with NCCL/RCCL
- [ ] Support for mixed GPU vendors (CUDA + HIP)
- [ ] Minimal communication overhead (< 10%)

### 1. Multi-GPU Architecture Design
- [ ] Design data parallelism strategy
- [ ] Implement GPU rank/world size management
- [ ] Setup inter-GPU communication topology
- [ ] Design gradient synchronization protocol
- [ ] Support heterogeneous GPU configurations

**Files**:
- `include/llm/lora_framework/multi_gpu.h` (new)
- `src/llm/lora_framework/multi_gpu.cpp` (new)

**Architecture**:
```
GPU 0 (Rank 0)     GPU 1 (Rank 1)     GPU 2 (Rank 2)     GPU 3 (Rank 3)
   ↓                   ↓                   ↓                   ↓
Data Split 0       Data Split 1       Data Split 2       Data Split 3
   ↓                   ↓                   ↓                   ↓
Forward Pass       Forward Pass       Forward Pass       Forward Pass
   ↓                   ↓                   ↓                   ↓
Backward Pass      Backward Pass      Backward Pass      Backward Pass
   ↓                   ↓                   ↓                   ↓
Local Gradients    Local Gradients    Local Gradients    Local Gradients
   └───────────────────┴───────────────────┴───────────────────┘
                            ↓
                    All-Reduce (NCCL/RCCL)
                            ↓
                    Average Gradients
                            ↓
   ┌───────────────────┬───────────────────┬───────────────────┐
   ↓                   ↓                   ↓                   ↓
Optimizer Step     Optimizer Step     Optimizer Step     Optimizer Step
```

### 2. NCCL Integration (NVIDIA)
- [ ] Initialize NCCL communicator
- [ ] Implement NCCL all-reduce for gradients
- [ ] Setup NCCL streams for async communication
- [ ] Support NCCL peer-to-peer transfers
- [ ] Implement NCCL error handling

**Files**:
- `include/llm/lora_framework/nccl_backend.h` (new)
- `src/llm/lora_framework/nccl_backend.cpp` (new)

**Dependencies**: NCCL 2.10+

### 3. RCCL Integration (AMD)
- [ ] Initialize RCCL communicator
- [ ] Implement RCCL all-reduce for gradients
- [ ] Setup RCCL streams for async communication
- [ ] Support RCCL peer-to-peer transfers
- [ ] Implement RCCL error handling

**Files**:
- `include/llm/lora_framework/rccl_backend.h` (new)
- `src/llm/lora_framework/rccl_backend.cpp` (new)

**Dependencies**: RCCL 2.10+

### 4. Custom All-Reduce Implementation (Fallback)
- [ ] Implement ring all-reduce algorithm
- [ ] Support for mixed GPU vendors (CUDA + HIP)
- [ ] Point-to-point GPU transfers
- [ ] Overlap communication with computation
- [ ] Fallback when NCCL/RCCL unavailable

**Files**:
- `include/llm/lora_framework/custom_allreduce.h` (new)
- `src/llm/lora_framework/custom_allreduce.cpp` (new)

**Ring All-Reduce**:
```
Step 1: GPU i sends to GPU (i+1) % N
Step 2: GPU i receives from GPU (i-1) % N
...
N-1 steps total for all-reduce
```

### 5. Multi-GPU LoRA Layer
- [ ] Create `MultiGPULoRALayer` wrapper
- [ ] Distribute model replicas across GPUs
- [ ] Shard data batches across GPUs
- [ ] Synchronize gradients after backward pass
- [ ] Average gradients before optimizer step
- [ ] Support gradient accumulation across micro-batches

**Files**:
- `include/llm/lora_framework/multi_gpu_lora_layer.h` (new)
- `src/llm/lora_framework/multi_gpu_lora_layer.cpp` (new)

### 6. Multi-GPU Trainer
- [ ] Create `MultiGPULoRATrainer` class
- [ ] Implement data parallel training loop
- [ ] Distribute batches to GPUs
- [ ] Synchronize after each training step
- [ ] Support gradient accumulation
- [ ] Implement distributed checkpointing

**Files**:
- `include/llm/lora_framework/multi_gpu_trainer.h` (new)
- `src/llm/lora_framework/multi_gpu_trainer.cpp` (new)

**API Example**:
```cpp
// Initialize multi-GPU context
MultiGPUContext ctx(num_gpus=4);

// Create multi-GPU layer
MultiGPULoRALayer layer(768, 768, 8, 1.0f, ctx);

// Distributed training
for (int epoch = 0; epoch < num_epochs; ++epoch) {
    // Data loader automatically shards batches
    for (auto& batch : distributed_data_loader) {
        // Each GPU processes its shard
        auto outputs = layer.forward(batch);
        
        // Compute local loss
        float local_loss = mse_loss(outputs, targets);
        
        // Backward on each GPU
        layer.backward(grad_loss);
        
        // Gradients synchronized automatically
        optimizer.step();  // All GPUs update with averaged gradients
    }
}
```

### 7. Data Distribution
- [ ] Implement distributed data loader
- [ ] Shard training data across GPUs
- [ ] Support distributed sampling
- [ ] Implement distributed batching
- [ ] Handle uneven data distribution

**Files**:
- `include/llm/lora_framework/distributed_dataloader.h` (new)
- `src/llm/lora_framework/distributed_dataloader.cpp` (new)

### 8. Communication Optimization
- [ ] Overlap gradient all-reduce with backward pass
- [ ] Use gradient bucketing for efficiency
- [ ] Implement gradient compression (optional)
- [ ] Optimize communication topology (NVLink-aware)
- [ ] Profile communication overhead

**Techniques**:
- Gradient bucketing: Group small tensors for fewer all-reduce calls
- Overlap: Start all-reduce as soon as gradients ready
- Compression: FP16 gradients reduce bandwidth

### 9. Testing and Validation
- [ ] Unit tests for multi-GPU primitives
- [ ] Test gradient synchronization correctness
- [ ] Numerical accuracy tests (single GPU vs multi-GPU)
- [ ] Performance scaling tests (1, 2, 4, 8 GPUs)
- [ ] Test on heterogeneous GPU configurations
- [ ] Communication overhead profiling

**Files**:
- `tests/test_multi_gpu.cpp` (new)
- `tests/test_distributed_training.cpp` (new)

### 10. Documentation
- [ ] Multi-GPU training guide
- [ ] Setup instructions for NCCL/RCCL
- [ ] Performance tuning for multi-GPU
- [ ] Troubleshooting communication issues
- [ ] Scaling efficiency analysis
- [ ] Update Phase 10 plan with results

## ✅ Acceptance Criteria

- [ ] Multi-GPU training functional on 2, 4, 8 GPUs
- [ ] Gradient synchronization correct (matches single GPU)
- [ ] **Performance**: Near-linear scaling (4 GPUs → 3.5-3.8x)
- [ ] **Accuracy**: No degradation vs single GPU
- [ ] **Overhead**: Communication < 10% of training time
- [ ] NCCL and RCCL backends functional
- [ ] Fallback all-reduce works without NCCL/RCCL
- [ ] Support for mixed GPU vendors
- [ ] All tests pass on multi-GPU systems

## 🔗 Dependencies

- NCCL 2.10+ (for NVIDIA multi-GPU)
- RCCL 2.10+ (for AMD multi-GPU)
- CUDA 11.8+ or ROCm 5.0+
- Multiple GPUs (2, 4, or 8 recommended)
- NVLink or PCIe interconnect
- Existing GPU backends (CUDA, HIP functional)

### Scaling Efficiency
| GPUs | Training Time | Speedup | Efficiency |
|------|---------------|---------|------------|
| 1 GPU | 3.2ms | 1.0x | 100% |
| 2 GPUs | 1.7ms | 1.88x | 94% |
| 4 GPUs | 0.9ms | 3.56x | 89% |
| 8 GPUs | 0.5ms | 6.40x | 80% |

**Target**: > 85% scaling efficiency for 4 GPUs

### Communication Overhead
| Component | Time (4 GPUs) | % of Total |
|-----------|---------------|------------|
| Forward | 0.25ms | 28% |
| Backward | 0.50ms | 55% |
| All-Reduce | 0.10ms | 11% |
| Optimizer | 0.05ms | 6% |
| **Total** | **0.90ms** | **100%** |

**Target**: All-reduce < 10% of total time

### Combined Performance (All Optimizations)
| Configuration | Training Step Time | Total Speedup |
|---------------|-------------------|---------------|
| CPU (baseline) | 160ms | 1x |
| 1 GPU (FP32) | 3.2ms | 50x |
| 1 GPU + Fusion | 1.75ms | 91x |
| 1 GPU + Fusion + FP16 | 0.9ms | 178x |
| **4 GPUs + All Optimizations** | **0.25ms** | **640x** |

### Memory Requirements (Per GPU)
| Model | Single GPU | 4 GPUs (Per GPU) | Total |
|-------|------------|------------------|-------|
| Llama-7B (FP16, r=8) | 5 GB | 5 GB | 20 GB |
| Llama-13B (FP16, r=8) | 8 GB | 8 GB | 32 GB |
| Llama-30B (FP16, r=8) | 16 GB | 16 GB | 64 GB |

**Note**: Each GPU holds full model replica but processes different data

## 📚 References

- NCCL Documentation: https://docs.nvidia.com/deeplearning/nccl/
- RCCL Documentation: https://rocm.docs.amd.com/projects/rccl/
- Distributed Training Guide: https://pytorch.org/tutorials/intermediate/ddp_tutorial.html
- Ring All-Reduce Algorithm: https://tech.preferred.jp/en/blog/technologies-behind-distributed-deep-learning-allreduce/
- Phase 10 Plan: `LORA_GPU_PHASE10_PLAN.md`

### Data Parallelism Strategy
1. **Model Replication**: Each GPU has full copy of model
2. **Data Sharding**: Training data split across GPUs
3. **Forward/Backward**: Independent computation on each GPU
4. **Gradient Sync**: All-reduce to average gradients
5. **Update**: All GPUs update with same averaged gradients

### NCCL All-Reduce Example
```cpp
// Initialize NCCL
ncclComm_t comm;
ncclCommInitRank(&comm, num_gpus, nccl_id, rank);

// All-reduce gradients
ncclAllReduce(
    grad_send_buff,      // Send buffer
    grad_recv_buff,      // Receive buffer
    grad_count,          // Number of elements
    ncclFloat,           // Data type
    ncclSum,             // Reduction operation
    comm,                // Communicator
    cuda_stream          // CUDA stream
);

// Average by number of GPUs
scalar_multiply_kernel(grad_recv_buff, 1.0f / num_gpus);
```

### Communication Overlap
```cpp
// Start backward pass
layer.backward(grad_output);

// As soon as each layer's gradient ready, start all-reduce
for (auto& param : layer.parameters()) {
    if (param.grad_ready()) {
        // Async all-reduce (non-blocking)
        nccl_all_reduce_async(param.grad, comm, stream);
    }
}

// Wait for all all-reduces to complete
ncclGroupEnd();
```

### Heterogeneous GPU Support
- Use custom all-reduce for mixed NVIDIA/AMD GPUs
- Fallback to CPU intermediate buffer if needed
- P2P transfers where available (same vendor)

---

**Status**: Not Started  
**Estimated Completion**: Week 15 of Phase 10  
**Expected Benefit**: 3.5-4x speedup with 4 GPUs (cumulative 178x → 640x vs CPU)  
**Enables**: Training larger models (Llama-65B) across multiple GPUs
